# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/cookie/CookieStorage.cpp
# Commit: 36aa6faea947
# Full Hash: 36aa6faea9477a13b6d94e396596697e56675018
# Author: Paul Zuehlcke <pbz@mozilla.com>
# Date: 2023-09-19 03:53:26
# Regressor Bug: 1839918
# File Overlap Count: 6
# Description:
#   Bug 1839918 - Refactor "cookie-changed" notifications to use nsICookieNotification. r=extension-reviewers,cookie-reviewers,sessionstore-reviewers,bvandersloot,edgul,robwu
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D185300
# ==============================================================================

diff -r 905592ffa91d -r 36aa6faea947 netwerk/cookie/CookieStorage.cpp
--- a/netwerk/cookie/CookieStorage.cpp	Mon Sep 18 11:25:24 2023 +0000
+++ b/netwerk/cookie/CookieStorage.cpp	Mon Sep 18 11:55:08 2023 +0000
@@ -6,6 +6,9 @@
 #include "Cookie.h"
 #include "CookieCommons.h"
 #include "CookieLogging.h"
+#include "CookieNotification.h"
+#include "nsCOMPtr.h"
+#include "nsICookieNotification.h"
 #include "CookieStorage.h"
 #include "mozilla/dom/nsMixedContentBlocker.h"
 #include "nsIMutableArray.h"
@@ -281,7 +284,7 @@
 
   if (cookie) {
     // Everything's done. Notify observers.
-    NotifyChanged(cookie, u"deleted");
+    NotifyChanged(cookie, nsICookieNotification::COOKIE_DELETED, aBaseDomain);
   }
 }
 
@@ -311,7 +314,8 @@
       RemoveCookieFromList(iter);
 
       if (cookie) {
-        NotifyChanged(cookie, u"deleted");
+        NotifyChanged(cookie, nsICookieNotification::COOKIE_DELETED,
+                      aBaseDomain);
       }
     }
   }
@@ -345,7 +349,8 @@
       RemoveCookieFromList(iter);
 
       if (cookie) {
-        NotifyChanged(cookie, u"deleted");
+        NotifyChanged(cookie, nsICookieNotification::COOKIE_DELETED,
+                      aBaseDomain);
       }
     }
   }
@@ -360,27 +365,42 @@
 
   RemoveAllInternal();
 
-  NotifyChanged(nullptr, u"cleared");
+  NotifyChanged(nullptr, nsICookieNotification::ALL_COOKIES_CLEARED, ""_ns);
 }
 
-// notify observers that the cookie list changed. there are five possible
-// values for aData:
-// "deleted" means a cookie was deleted. aSubject is the deleted cookie.
-// "added"   means a cookie was added. aSubject is the added cookie.
-// "changed" means a cookie was altered. aSubject is the new cookie.
-// "cleared" means the entire cookie list was cleared. aSubject is null.
-// "batch-deleted" means a set of cookies was purged. aSubject is the list of
-// cookies.
-void CookieStorage::NotifyChanged(nsISupports* aSubject, const char16_t* aData,
+// notify observers that the cookie list changed.
+void CookieStorage::NotifyChanged(nsISupports* aSubject,
+                                  nsICookieNotification::Action aAction,
+                                  const nsACString& aBaseDomain,
+                                  dom::BrowsingContext* aBrowsingContext,
+                                  bool aIsThirdPartyCookie,
                                   bool aOldCookieIsSession) {
   nsCOMPtr<nsIObserverService> os = services::GetObserverService();
   if (!os) {
     return;
   }
+
+  nsCOMPtr<nsICookie> cookie;
+  nsCOMPtr<nsIArray> batchDeletedCookies;
+
+  if (aAction == nsICookieNotification::COOKIES_BATCH_DELETED) {
+    batchDeletedCookies = do_QueryInterface(aSubject);
+  } else {
+    cookie = do_QueryInterface(aSubject);
+  }
+
+  uint64_t browsingContextId = 0;
+  if (aBrowsingContext) {
+    browsingContextId = aBrowsingContext->Id();
+  }
+
+  nsCOMPtr<nsICookieNotification> notification =
+      new CookieNotification(aAction, cookie, aBaseDomain, batchDeletedCookies,
+                             browsingContextId, aIsThirdPartyCookie);
   // Notify for topic "private-cookie-changed" or "cookie-changed"
-  os->NotifyObservers(aSubject, NotificationTopic(), aData);
+  os->NotifyObservers(notification, NotificationTopic(), u"");
 
-  NotifyChangedInternal(aSubject, aData, aOldCookieIsSession);
+  NotifyChangedInternal(notification, aOldCookieIsSession);
 }
 
 // this is a backend function for adding a cookie to the list, via SetCookie.
@@ -393,7 +413,9 @@
                               const OriginAttributes& aOriginAttributes,
                               Cookie* aCookie, int64_t aCurrentTimeInUsec,
                               nsIURI* aHostURI, const nsACString& aCookieHeader,
-                              bool aFromHttp) {
+                              bool aFromHttp,
+                              dom::BrowsingContext* aBrowsingContext,
+                              bool aIsThirdPartyCookie) {
   int64_t currentTime = aCurrentTimeInUsec / PR_USEC_PER_SEC;
 
   CookieListIter exactIter{};
@@ -513,7 +535,9 @@
       if (aCookie->Expiry() <= currentTime) {
         COOKIE_LOGFAILURE(SET_COOKIE, aHostURI, aCookieHeader,
                           "previously stored cookie was deleted");
-        NotifyChanged(oldCookie, u"deleted", oldCookieIsSession);
+        NotifyChanged(oldCookie, nsICookieNotification::COOKIE_DELETED,
+                      aBaseDomain, aBrowsingContext, aIsThirdPartyCookie,
+                      oldCookieIsSession);
         return;
       }
 
@@ -590,10 +614,15 @@
   // Now that list mutations are complete, notify observers. We do it here
   // because observers may themselves attempt to mutate the list.
   if (purgedList) {
-    NotifyChanged(purgedList, u"batch-deleted");
+    NotifyChanged(purgedList, nsICookieNotification::COOKIES_BATCH_DELETED,
+                  ""_ns);
   }
 
-  NotifyChanged(aCookie, foundCookie ? u"changed" : u"added",
+  // Notify for topic "private-cookie-changed" or "cookie-changed"
+  NotifyChanged(aCookie,
+                foundCookie ? nsICookieNotification::COOKIE_CHANGED
+                            : nsICookieNotification::COOKIE_ADDED,
+                aBaseDomain, aBrowsingContext, aIsThirdPartyCookie,
                 oldCookieIsSession);
 }
 